Spigot-Server-Patches/0435-Pillager-patrol-spawn-settings-and-per-player-option.patch

```diff
@@ -0,0 +1,144 @@
+ 
+ public class MobSpawnerPatrol {
+ 
++    private int getSpawnDelay() { return a; } // Paper - OBFHELPER
++    private void setSpawnDelay(int spawnDelay) { this.a = spawnDelay; } // Paper - OBFHELPER
+     private int a;
+ 
+     public MobSpawnerPatrol() {}
+ 
+     public int a(WorldServer worldserver, boolean flag, boolean flag1) {
+-        if (worldserver.paperConfig.disablePillagerPatrols) return 0; // Paper
++        if (worldserver.paperConfig.disablePillagerPatrols || worldserver.paperConfig.patrolSpawnChance == 0) return 0; // Paper
+         if (!flag) {
+             return 0;
+         } else if (!worldserver.getGameRules().getBoolean(GameRules.DO_PATROL_SPAWNING)) {
+@@ -17,23 +19,51 @@ public class MobSpawnerPatrol {
+         } else {
+             Random random = worldserver.random;
+ 
+-            --this.a;
+-            if (this.a > 0) {
++            // Paper start - Patrol settings
++            // Random player selection moved up for per player spawning and configuration
++            int j = worldserver.getPlayers().size();
++            if (j < 1) {
+                 return 0;
++            }
++
++            EntityPlayer entityhuman = worldserver.getPlayers().get(random.nextInt(j));
++            if (entityhuman.isSpectator()) {
++                return 0;
++            }
++
++            int patrolSpawnDelay;
++            if (worldserver.paperConfig.patrolPerPlayerDelay) {
++                --entityhuman.patrolSpawnDelay;
++                patrolSpawnDelay = entityhuman.patrolSpawnDelay;
+             } else {
+-                this.a += 12000 + random.nextInt(1200);
+-                long i = worldserver.getDayTime() / 24000L;
++                setSpawnDelay(getSpawnDelay() - 1);
++                patrolSpawnDelay = getSpawnDelay();
++            }
++
++            if (patrolSpawnDelay > 0) {
++                return 0;
++            } else {
++                long days;
++                if (worldserver.paperConfig.patrolPerPlayerStart) {
++                    days = entityhuman.getStatisticManager().getStatisticValue(StatisticList.CUSTOM.get(StatisticList.PLAY_ONE_MINUTE)) / 24000L; // PLAY_ONE_MINUTE is actually counting in ticks, a misnomer by Mojang
++                } else {
++                    days = worldserver.getDayTime() / 24000L;
++                }
++                if (worldserver.paperConfig.patrolPerPlayerDelay) {
++                    entityhuman.patrolSpawnDelay += worldserver.paperConfig.patrolDelay + random.nextInt(1200);
++                } else {
++                    setSpawnDelay(getSpawnDelay() + worldserver.paperConfig.patrolDelay + random.nextInt(1200));
++                }
+ 
+-                if (i >= 5L && worldserver.isDay()) {
+-                    if (random.nextInt(5) != 0) {
++                if (days >= worldserver.paperConfig.patrolStartDay && worldserver.isDay()) {
++                    if (random.nextDouble() >= worldserver.paperConfig.patrolSpawnChance) {
```

Comment:
not sure if it's my head spinning or if I'm actually brain here, but is this the correct way around?

Reply:
It is: If the random value is equal or above the spawn chance it will not attempt to spawn. E.g. a spawn chance of 0.2 and a random value of 0.6 would not result in a spawn. (I know it's a bit unintuitive but it has to be this way around because of the structure of the vanilla code which checks if nextInt(5) is 0 so anything above 0 would not hit it)